Readme.md
# Notification

  Notification component with a clean slate to build off of,
  style and position them however you like.

  ![js growl component](http://f.cl.ly/items/143P3r3X1E1m0Y0a2E2l/Screen%20Shot%202012-07-26%20at%203.16.11%20PM.png)

  These don't _have_ to look like growl style notifications, use
  your trusty friend CSS.

## Installation

```
$ npm install notification-component
```

## Features

  - events for composition
  - structural CSS letting you decide on style
  - transient notifications
  - transient closable notifications
  - sticky (implicitly closable) notifications
  - notification classes (info, warn, error)
  - fluent API

## Events

  - `close` the notification is closed via the X
  - `click` the notification is clicked

## API

### notify(msg)

  Notify with the given `msg` and no title. The
  notification will hide after 4 seconds by default.

### notify(title, msg)

  Notify with the given `msg` and `title`. The
  notification will hide after 4 seconds by default.

### notify.info(title, [msg])

  Same as `notify()`

### notify.warn(title, [msg])

  Same as `notify()` with a `warn` class for styling.

### notify.error(title, [msg])

  Same as `notify()` with a `error` class for styling.

### Notification#sticky()

  Make the notification sticky, aka it will not close
  automatically, and it will automatically be `.closable()`.

### Notification#show()

  Show the notification.

### Notification#hide([ms])

  Hide the notification immediately or wait ms.

### Notification#closable()

  Mark the notification as closable, adding an "X" so the user
  may explicitly close it.

### Notification#effect(name)

  One of the following effects, or define your own with class `name`:

  - `slide`
  - `fade`
  - `scale`

## License

  MIT


History.md

0.0.7 / 2015-07-21
==================

  * component: update "dom" to v1.0.8
  * docs: add note about optional ms for `hide()`

0.0.6 / 2015-03-11
==================

  * re-add the `package.json` file
  * component: update "dom" to v1.0.7
  * component: update "emitter" to v1.2.0
  * component: update "description"
  * component: Pin deps (#10)
  * More readable HTML entity for the close button

0.0.4 / 2014-03-17
==================

  * Use escaped "×"
  * switching over to component/dom (as per @ianstormtaylor's comment)
  * remove jquery dependency
  * add test command to Makefile
  * Update emitter to 1.0.1
  * add .license property to component.json
  * add private .classname option

0.0.3 / 2012-08-30
==================

  * add "close" event
  * add `.Notification` export


index.js

/**
 * Module dependencies.
 */

var dom = require('dom');
var Emitter = require('emitter');
var onBody = require('on-body');

/**
 * Expose `notify`.
 */

exports = module.exports = notify;

/**
 * Notification list.
 */

var list = dom('<ul id="notifications">');

/**
 * Append to body when it exists.
 */

onBody(function (body) {
  list.appendTo(body);
});

/**
 * Return a new `Notification` with the given
 * (optional) `title` and `msg`.
 *
 * @param {String} title or msg
 * @param {String} msg
 * @return {Dialog}
 * @api public
 */

function notify(title, msg){
  switch (arguments.length) {
    case 2:
      return new Notification({ title: title, message: msg })
        .show()
        .hide(4000);
    case 1:
      return new Notification({ message: title })
        .show()
        .hide(4000);
  }
}

/**
 * Construct a notification function for `type`.
 *
 * @param {String} type
 * @return {Function}
 * @api private
 */

function type(type) {
  return function(title, msg){
    return notify.apply(this, arguments)
      .type(type);
  };
}

/**
 * Notification methods.
 */

exports.info = notify;
exports.warn = type('warn');
exports.error = type('error');

/**
 * Expose constructor.
 */

exports.Notification = Notification;

/**
 * Initialize a new `Notification`.
 *
 * Options:
 *
 *    - `title` dialog title
 *    - `message` a message to display
 *
 * @param {Object} options
 * @api public
 */

function Notification(options) {
  Emitter.call(this);
  options = options || {};
  this.el = dom(require('./template'));
  this.render(options);
  if (options.classname) this.el.addClass(options.classname);
  if (Notification.effect) this.effect(Notification.effect);
}

/**
 * Inherit from `Emitter.prototype`.
 */

Notification.prototype = new Emitter;

/**
 * Render with the given `options`.
 *
 * @param {Object} options
 * @api public
 */

Notification.prototype.render = function(options){
  var el = this.el
    , title = options.title
    , msg = options.message
    , self = this;

  el.find('.close').on('click', function(){
    self.emit('close');
    self.hide();
    return false;
  });

  el.on('click', function(e){
    e.preventDefault();
    self.emit('click', e);
  });

  el.find('.title').text(title);
  if (!title) el.find('.title').remove();

  // message
  if ('string' == typeof msg) {
    el.find('p').text(msg);
  } else if (msg) {
    el.find('p').replace(msg.el || msg);
  }

  setTimeout(function(){
    el.removeClass('hide');
  }, 0);
};

/**
 * Enable the dialog close link.
 *
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.closable = function(){
  this.el.addClass('closable');
  return this;
};

/**
 * Set the effect to `type`.
 *
 * @param {String} type
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.effect = function(type){
  this._effect = type;
  this.el.addClass(type);
  return this;
};

/**
 * Show the notification.
 *
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.show = function(){
  this.el.appendTo(list);
  return this;
};

/**
 * Set the notification `type`.
 *
 * @param {String} type
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.type = function(type){
  this._type = type;
  this.el.addClass(type);
  return this;
};

/**
 * Make it stick (clear hide timer), and make it closable.
 *
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.sticky = function(){
  return this.hide(0).closable();
};

/**
 * Hide the dialog with optional delay of `ms`,
 * otherwise the notification is removed immediately.
 *
 * @return {Number} ms
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.hide = function(ms){
  var self = this;

  // duration
  if ('number' == typeof ms) {
    clearTimeout(this.timer);
    if (!ms) return this;
    this.timer = setTimeout(function(){
      self.hide();
    }, ms);
    return this;
  }

  // hide / remove
  this.el.addClass('hide');
  if (this._effect) {
    setTimeout(function(self){
      self.remove();
    }, 500, this);
  } else {
    self.remove();
  }

  return this;
};

/**
 * Hide the notification without potential animation.
 *
 * @return {Dialog} for chaining
 * @api public
 */

Notification.prototype.remove = function(){
  this.el.remove();
  return this;
};

notification.js

/**
 * Module dependencies.
 */

var Emitter = require('emitter')
  , o = require('jquery');

/**
 * Notification list.
 */

var list;

/**
 * Expose `notify`.
 */

exports = module.exports = notify;

// list

o(function(){
  list = o('<ul id="notifications">');
  list.appendTo('body');
})

/**
 * Return a new `Notification` with the given 
 * (optional) `title` and `msg`.
 *
 * @param {String} title or msg
 * @param {String} msg
 * @return {Dialog}
 * @api public
 */

function notify(title, msg){
  switch (arguments.length) {
    case 2:
      return new Notification({ title: title, message: msg })
        .show()
        .hide(4000);
    case 1:
      return new Notification({ message: title })
        .show()
        .hide(4000);
  }
}

/**
 * Construct a notification function for `type`.
 *
 * @param {String} type
 * @return {Function}
 * @api private
 */

function type(type) {
  return function(title, msg){
    return notify.apply(this, arguments)
      .type(type);
  }
}

/**
 * Notification methods.
 */

exports.info = notify;
exports.warn = type('warn');
exports.error = type('error');

/**
 * Initialize a new `Notification`.
 *
 * Options:
 *
 *    - `title` dialog title
 *    - `message` a message to display
 *
 * @param {Object} options
 * @api public
 */

function Notification(options) {
  Emitter.call(this);
  options = options || {};
  this.el = o(render('notification'));
  this.render(options);
  if (Notification.effect) this.effect(Notification.effect);
};

/**
 * Inherit from `Emitter.prototype`.
 */

Notification.prototype = new Emitter;

/**
 * Render with the given `options`.
 *
 * @param {Object} options
 * @api public
 */

Notification.prototype.render = function(options){
  var el = this.el
    , title = options.title
    , msg = options.message
    , self = this;

  el.find('.close').click(function(){
    self.hide();
    return false;
  });

  el.click(function(e){
    e.preventDefault();
    self.emit('click', e);
  });

  el.find('.title').text(title);
  if (!title) el.find('.title').remove();

  // message
  if ('string' == typeof msg) {
    el.find('p').text(msg);
  } else if (msg) {
    el.find('p').replaceWith(msg.el || msg);
  }

  setTimeout(function(){
    el.removeClass('hide');
  }, 0);
};

/**
 * Enable the dialog close link.
 *
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.closable = function(){
  this.el.addClass('closable');
  return this;
};

/**
 * Set the effect to `type`.
 *
 * @param {String} type
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.effect = function(type){
  this._effect = type;
  this.el.addClass(type);
  return this;
};

/**
 * Show the notification.
 *
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.show = function(){
  this.el.appendTo(list);
  return this;
};

/**
 * Set the notification `type`.
 *
 * @param {String} type
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.type = function(type){
  this._type = type;
  this.el.addClass(type);
  return this;
};

/**
 * Make it stick (clear hide timer), and make it closable.
 *
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.sticky = function(){
  return this.hide(0).closable();
};

/**
 * Hide the dialog with optional delay of `ms`,
 * otherwise the notification is removed immediately.
 *
 * @return {Number} ms
 * @return {Notification} for chaining
 * @api public
 */

Notification.prototype.hide = function(ms){
  var self = this;

  // duration
  if ('number' == typeof ms) {
    clearTimeout(this.timer);
    if (!ms) return this;
    this.timer = setTimeout(function(){
      self.hide();
    }, ms);
    return this;
  }

  // hide / remove
  this.el.addClass('hide');
  if (this._effect) {
    setTimeout(function(self){
      self.remove();
    }, 500, this);
  } else {
    self.remove();
  }

  return this;
};

/**
 * Hide the notification without potential animation.
 *
 * @return {Dialog} for chaining
 * @api public
 */

Notification.prototype.remove = function(){
  this.el.remove();
  return this;
};